### ISSUE TITLE
status_payload_printer.cc is not compiled with Clang 10 on Windows

### ISSUE THREAD

**Issue by samolisov** [abseil/abseil-cpp#659]
The following code in the `absl/status/status_payload_printer.cc` raises a compilation error on Windows when Clang 10 (via clang-cl.exe) is used to compile Abseil:

```c++
namespace {
// Tried constant initialized global variable but it doesn't work with Lexan
// (MSVC's `std::atomic` has trouble constant initializing).
std::atomic<StatusPayloadPrinter>& GetStatusPayloadPrinterStorage() {
  ABSL_CONST_INIT static std::atomic<StatusPayloadPrinter> instance{nullptr};
  return instance;
}
}  // namespace
```

The error:

```
..\absl\status\status_payload_printer.cc(28,60): error: variable does not have a constant initializer
  ABSL_CONST_INIT static std::atomic<StatusPayloadPrinter> instance{nullptr};
                                                           ^~~~~~~~~~~~~~~~~
..\absl\status\status_payload_printer.cc(28,3): note: required by 'require_constant_initialization' attribute here
  ABSL_CONST_INIT static std::atomic<StatusPayloadPrinter> instance{nullptr};
  ^~~~~~~~~~~~~~~
..\absl/base/attributes.h(618,27): note: expanded from macro 'ABSL_CONST_INIT'
#define ABSL_CONST_INIT [[clang::require_constant_initialization]]
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.15.26726\include\xxatomic(207,21): note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
                : _Atomic_address{(_ATOMIC_UINT)_Right}
                                  ^
..\absl\status\status_payload_printer.cc(28,60): note: in call to 'atomic(nullptr)'
  ABSL_CONST_INIT static std::atomic<StatusPayloadPrinter> instance{nullptr};
                                                           ^
1 error generated.
```

I see the following solution: to add a check the compiler is not MSVC (since clang-cl mimics MSVC) into the `ABSL_CONST_INIT` macro but it won't protect from the error if `clang++` is used instead of `clang-cl`. Maybe someone knows a better solution?

**Comment by zhangxy988** [comment]
The root cause isn't `ABSL_CONST_INIT` but MSVC's `std::atomic` implementation. So I think it's probably OK to just remove `ABSL_CONST_INIT` when compiled with MSVC:
```
#ifdef _MSC_VER
static std::atomic<StatusPayloadPrinter> instance{nullptr};
#else
ABSL_CONST_INIT static std::atomic<StatusPayloadPrinter> instance{nullptr};
#endif
```
Could you please try that? We currently don't have an environment with clang10+msvc.

**Comment by derekmauro** [comment]
https://github.com/abseil/abseil-cpp/blob/master/absl/base/internal/atomic_hook.h exists specifically to work around this issue.

**Comment by samolisov** [comment]
@derekmauro I've seen into `atomic_hook.h` and see the proposed `ABSL_HAVE_WORKING_CONSTEXPR_STATIC_INIT` is defined as the following:

```c++
#if defined(_MSC_VER) && !defined(__clang__)
#define ABSL_HAVE_WORKING_CONSTEXPR_STATIC_INIT 0
#else
#define ABSL_HAVE_WORKING_CONSTEXPR_STATIC_INIT 1
#endif
```

But `clang-cl` defines both `_MSC_VER` and `__clang__` and the macro expanded to 1:

```c++
..\absl/base/internal/atomic_hook.h(49,46): note: expanded from macro 'ABSL_INTERNAL_ATOMIC_HOOK_ATTRIBUTES'
#define ABSL_INTERNAL_ATOMIC_HOOK_ATTRIBUTES ABSL_CONST_INIT
                                             ^~~~~~~~~~~~~~~
..\absl/base/attributes.h(618,27): note: expanded from macro 'ABSL_CONST_INIT'
#define ABSL_CONST_INIT [[clang::require_constant_initialization]]
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I would propose just to remove the `!defined(__clang__)` condition, the same way as `ABSL_HAVE_WORKING_ATOMIC_POINTER` defined but as I see, `__clang__` check was added in commit 08a7e7bf - move from internal Abseil repo, where this change is described as:

```
d6916d30c5c1d3ee9ae46d69ec0a166a760c99c7 by Derek Mauro <dmauro@google.com>:

Make AtomicHook constant-initializable on Clang for Windows.

Only mark AtomicHook as constant-initializable on platforms where it
is actually constant-initializable.
```

**Comment by derekmauro** [comment]
@samolisov I think you are misinterpreting my suggestion. I'm suggesting that the variable that holds `StatusPayloadPrinter` be `absl::base_internal::AtomicHook<StatusPayloadPrinter>` instead of `std::atomic<StatusPayloadPrinter>`.

**Comment by zhangxy988** [comment]
@samolisov Could you please try the above approach suggested by @derekmauro and send us a pull request, i.e. replace `std::atomic<StatusPayloadPrinter>` with `absl::base_internal::AtomicHook<StatusPayloadPrinter>`? Thanks!

**Comment by samolisov** [comment]
@derekmauro I'm sorry, yes it looks I misinterpreted your answer. Thank you for your suggestion about `AtomicHook`.

@zhangxy988 I've updated `status_payload_printer.cc` to use `AtomicHook` instead of plain `std::atomic`. See #661. One difference: before the functions `SetStatusPayloadPrinter` and `GetStatusPayloadPrinter` used `load(..., std::memory_order_relaxed)` and `store(..., std::memory_order_relaxed)` while `AtomicHook` uses `std::memory_order_acquire` to load and store. Is there any benchmark to check how it affect the performance?
